app/scripts/services/authorization.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("AuthorizationService", function($q, $cacheFactory, Logger, $interval, DataService){
+    function AuthorizationService() {
+      this.currentProject = null;
+      this.cachedRulesByProject = $cacheFactory('rulesCache', {
+        number: 10
+      });
+    }
+
+    // Transform data from:
+    // rules = {resources: ["jobs"], apiGroups: ["extensions"], verbs:["create","delete","get","list","update"]}
+    // into:
+    // normalizedRules = {"extensions/jobs": ["create","delete","get","list","update"]}
+    // Also while looping through the rules, check if user can create/update any resource on the 'Add to project' so the button will be displayed.
+    var normalizeRules = function(rules) {
+      var normalizedRules = {};
+      var canAddToProject = null;
+
+      // Check if resource name meets one of following conditions, since those resources can't be create/update via `Add to project` page:
+      //  - 'projectrequests'
+      //  - subresource taht contains '/', eg: 'builds/source', 'builds/logs', ...
+      //  - resource ending with 'review' string: 'localsubjectaccessreviews', 'selfsubjectrulesreviews', 'subjectaccessreviews' 
+      var checkResource = function(resource) {
+        if (resource === "projectrequests" || _.contains(resource, "/") || resource.endsWith("reviews")) {
+          return false;
+        } else {
+          return true;
+        }
+      };
+      _.each(rules, function(rule) {
+        _.each(rule.resources, function(resource) {
+          _.each(rule.apiGroups, function(apiGroup) {
+            normalizedRules[(apiGroup === "") ? resource : apiGroup + "/" + resource] = rule.verbs;
+          });
+
+          if (!canAddToProject && checkResource(resource) && (!_.isEmpty(_.intersection(rule.verbs ,(["*", "create", "update"]))))) {
+            canAddToProject = true;
+          }
+        });
+      });
+      return [normalizedRules, canAddToProject];
+    };
+
+    AuthorizationService.prototype.getProjectRules = function(projectName) {
+      var self = this;
+      var deferred = $q.defer();
```

Comment:
Probably don't need to create a generic deferred here, I would suggest the function return the `DataService.create` (which is a promise), or at least `deferred.resolve(projectRules)` so the resolution is data.  

I believe you can do something like:

```javascript
// return the promise
return DataService
            .create('selfsubjectrulesreviews', / *  args */)
            // tap in here with .then() to do your intermediate work
            .then(function resolve(data) {
              self.cachedRulesByProject.set(projectName, /* stuff */);
              return self.cachedRulesByProject.get(projectName);  // nice to get it from my call to getProjectRules()
            }, function reject(err) {
               // do err handling if you want... or don't and let calling code deal with this
               return err;
            });

```